### ISSUE TITLE
go/types: missing Info.Types for *ast.FuncType

### ISSUE THREAD

**Issue by adonovan** [golang/go#70908]
In this program, types.Info.Types records no type for f's ast.FuncType node. I think it should:
```go
package p

func f() (int, string) {
    return 0, "" 
}
```
<img width="333" alt="Screenshot 2024-12-18 at 1 29 13 PM" src="https://github.com/user-attachments/assets/a1a51e8f-2507-4ba1-a9d4-61f3742fcd93" />

@madelinekalil

**Comment by Agent-Hellboy** [comment]
Hi @adonovan,

I’d like to work on this if @madelinekalil  approve that capturing this information is the right approach. I’ve recently added  a PR on a mini parser for go.mod, and the Go codebase looks clean and approachable. I am particularly interested in the parser, AST, and type-checking stages, and I want to contribute to these areas as they seem beginner-friendly and a great way to start learning a language's type system through hands-on experience.

**Comment by adonovan** [comment]
> I’d like to work on this if @madelinekalil approve the issue. I’ve recently added a PR on a mini parser for go.mod, and the Go codebase looks clean and approachable. I am particularly interested in the parser, AST, and type-checking stages

Thanks for your interest!  @madelinekalil discovered the bug, but I think @griesemer is the best person to decide whether the simple fix you suggest is the right one. I know in the past there have been several discussions about when exactly the type checker guarantees to record a type for an expression, and it's possible this one lies in a grey area since, in context, the FuncType it isn't strictly a term or a type. Perhaps the most significant part of the fix will be to clarify the documentation on precisely what the user can safely assume.

> they seem beginner-friendly

I have never heard go/types described that way before! ;-)

**Comment by Agent-Hellboy** [comment]
> but I think @griesemer is the best person to decide whether the simple fix you suggest is the right one.

no problem. I can understand, I will keep looking at the issues and PRs and learning. 

> I have never heard go/types described that way before! ;-)

No no, I mean I am interested, and at least I can understand issues associated with these layers.

**Comment by griesemer** [comment]
I looked into this in fair detail:

For this program:
```Go
package p

var x int
var y = 1.0

func f() (int, string) {
    return 0, ""
}

var g func()
```
we get the following type recordings ([playground](https://go.dev/play/p/hHqcOiJcEbt?v=gotip)):
```
Types
	 p:7:12 	 0 	 {4 int 0}
	 p:7:15 	 "" 	 {4 string ""}
	 p:3:7 	 int 	 {3 int <nil>}
	 p:4:9 	 1.0 	 {4 float64 1}
	 p:6:11 	 int 	 {3 int <nil>}
	 p:6:16 	 string 	 {3 string <nil>}
	 p:10:7 	 func() 	 {3 func() <nil>}
Defs
	 p:10:5 	 g 	 var p.g func()
	 p:1:9 	 p 	 <nil>
	 p:3:5 	 x 	 var p.x int
	 p:4:5 	 y 	 var p.y float64
	 p:6:6 	 f 	 func p.f() (int, string)
```
The type for `f` is recorded (via an Object) in the `Info.Defs` map since `f` is a declaration, so the type information is definitely present. For the variable `x` we record the type in `Info.Defs` (as expected), but we also have the type `int` recorded in the `Info.Types` map since we have a (syntactic) type expression `int`.

Presumably, the argument is that since we record the type for the type expression `int` we should do the same for the signature "expressions" in function declarations.

Yet, we don't do it for function declarations, because there's no obvious _syntactic_ function type _expression_ present in the source (e.g. we don't write `f func() (int string)`), we mix the function type into the declaration. The `ast.FuncDecl` in th AST is artificial in that sense.

Note that we definitely record the function type in `info.Types` when we have an explicit _syntactic_ function type (line 10).

I am not convinced we should make the suggested change. For one it has been working like this for more than a decade and nobody complained. Second, I am concerned that recording this information may cause unexpected bugs (due to tools not expecting the data). Finally, for generic functions, `go/ast` records type parameters with the `ast.FuncType`, but the `syntax` package records type parameters with the `FuncDecl` node, so there's no obvious 1:1 correspondence between a `FuncType` node and the respective `types.Signature` in that sense, exactly because syntactically we actually don't have a clear-cut function type for function declarations.

As an aside, the [suggested fix](https://github.com/golang/go/issues/70908#issuecomment-2559427045) doesn't look right; that type information is recorded externally, and doing it in `Checker.funcType` may lead to repeated recordings.

All in all, I believe we should leave things as they are.
@adonovan, @findleyr Thoughts?

**Comment by adonovan** [comment]
Yeah, I agree it's a grey area. I think a reasonable resolution would be to document more explicitly at Info.Types what the client can expect.
